### ISSUE TITLE
Android 64 bit Support needs to be enabled by default

### ISSUE THREAD

**Issue by NathanaelA** [NativeScript/nativescript-cli#4330]
**Describe the issue**
By [August 1st, Google](https://www.google.com/appserve/mkt/p/AHANi0akXIxL8GvvvN82ScceqNlEYCN2ca8eTa-FhECpK8HrWq-SBPq5yXhCFdyv4dL5kzSHOgjCX-NLP9Ep9NnZfHzUuEpnFtatGVlzoNgyIcBrypwYA-QRjKA46VYUcbGCXyfp-Pfco0COy6KNactY8D-MH_z1zFTf
) will require apps to have 64 bit support enabled in addition to the 32 bit support.    

Because the actual nativescript runtimes are native code (i.e. c/c++) this is a requirement we have to do, so getting this re-enabled by default; sooner than later will simplify everyone's life before the Aug 1st deadline.  

**Additional context**
The 64 bit runtimes are shipped, but in the past were disabled by default to keep the .apk size smaller.  
Indirectly related to #1392

**Comment by rosen-vladimirov** [comment]
Hey @NathanaelA ,
Thanks for bringing the attention to this new requirement. We'll discuss it internally and come up with a suggestion how to support this. It looks like we'll have to enable builds for both armv7 and arm64 architectures by default. This will make the build slower, so we may enable it for release builds only. At the moment users can control the architectures from the `app.gradle` file in the project, but we may add some command line flag to make it easier. What do you think for all of these ideas?

**Comment by NathanaelA** [comment]
@rosen-vladimirov - I think enabling it only on release is the best idea.   Adding flags would be useful for people wanting to test; but as long as we document how to change the app.gradle for debug builds I think your good.   ;-)

**Comment by brndusic** [comment]
I added this to my app.gradle and it worked fine.

```
defaultConfig {
       ......
       ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86','x86_64'
}
```

**Comment by racknoris** [comment]
@brndusic Did you have any troubles in google play? Did the added 64 bit support affect any plugins?

@rosen-vladimirov Is this line what you were planning to add when using a flag?

**Comment by brndusic** [comment]
@racknoris I haven't noticed any problem. Warning on google play went away after I built this way.

**Comment by nlmunro** [comment]
I can confirm this solved it for me - using geolocation, firebase, localstorage and som UI plugins.

**Comment by racknoris** [comment]
@brndusic It seems that [this](https://docs.nativescript.org/angular/tooling/publishing/android-app-bundle#available-configurations) is the official way to do it, does it do the same as your way?
```
android {
....
  defaultConfig {
    ....
    ndk {
      abiFilters.clear()
    }
  }
....
```

**Comment by lambourn** [comment]
Maybe I'm missing something here - I just built the Android version of our app and set the app.gradle as suggested by @brndusic 
```
defaultConfig {
       ......
       ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86','x86_64'
}
```
however, running APK Analyze from Android studio I do not see libNativeScript.so for x86_64:

![apk-analyzer](https://user-images.githubusercontent.com/20942219/58466511-12805f80-813a-11e9-823a-1763622e7837.png)

**Comment by racknoris** [comment]
@lambourn could you please try in the way I found as well? perhaps it'll work

**Comment by lambourn** [comment]
@racknoris I did. Same result.

**Comment by brndusic** [comment]
@racknoris I have not analyzed .apk file, the thing it solved for me is that there is no more warning on app store about 64 bit support.

**Comment by lambourn** [comment]
@brndusic good to know. Maybe the Play Store is fine with just the arm64 binaries being present as x86_64 is not a relevant platform for Android mobile phones / tablets?

**Comment by ExaltedJim** [comment]
I am experiencing a crash of my app on startup when running it on an Android emulator with the CPU set to x86_64. This is because libNativeScript.so is missing from lib/x86_64 folder of the apk (it is present for all other architectures - x86, arm64-v8a, armeabi-v7a). Is there a plan to have a x86_64 version of libNativeScript.so?

**Comment by felix-idf** [comment]
@ExaltedJim ,
do you see any crash information in the logs?

**Comment by romandragan** [comment]
@ExaltedJim, same for me.

**Comment by kzimny** [comment]
My app.gradle looks as follow:
```
android {
  defaultConfig {  
    generatedDensities = []
	  applicationId = "com.aaaaa.nativeapp"
    ndk {
      abiFilters.clear()
    }
  }  
  aaptOptions {  
    additionalParameters "--no-version-vectors"  
  }  
} 
```
.apk package contains lib\arm64-v8a; lib\armeabi-v7a and lib\x86 folders with libNativeScript.so library
![image](https://user-images.githubusercontent.com/12007674/60988038-20352180-a343-11e9-903d-9e34c332357a.png)

Where is the libs\x86_64 folder? Do I miss something or it is expected?

**Comment by Serge-SDL** [comment]
same problem here and the deadline to update app is august first!

**Comment by dmdnkv** [comment]
The same problem. 
It was decided to skip x86 platform support for our app as a workaround since almost 99% of mobile devices use ARM.
The following config could help:
```android {
 defaultConfig {
   generatedDensities = []
   ndk {
       abiFilters.clear()
       abiFilters.addAll(['armeabi-v7a','arm64-v8a'])
   }
 }
 aaptOptions {
   additionalParameters "--no-version-vectors"
 }
}
```
Supported count of devices was decreased by 6 from 10557 to 10551 (that was displayed in Google Console App releases section after the roll out with that configuration)

**Comment by kzimny** [comment]
Thanks @dmdnkv, indeed not many mobile devices runs on x86 or x86_64 architecture. Found an interesting description [here](https://android.stackexchange.com/questions/186334/what-percentage-of-android-devices-runs-on-x86-architecture). I'll try with 'armeabi-v7a' and 'arm64-v8a' CPU's only.

**Comment by lambourn** [comment]
I just built an app with `ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86'`, exported as .aab. and uploaded it to the Play Store just fine, no complaints about missing 64-bit binaries. 

My assumption is that Google does not require x86_64 binaries at all (and if the app is really running on a x86_64 target, e.g. tablet. then the x86 32-bit binaries will be used)

**Comment by dmdnkv** [comment]
@lambourn according to this guide Google provided https://developer.android.com/distribute/best-practices/develop/64-bit seems to be that x86_64 is required. There were no complaints about missing x86_64 in Google Console for app I am working on and it is possible to upload the APK without 64bit version but the warning came from Google by email that app doesn't meet the requirement.

![image](https://user-images.githubusercontent.com/2073188/61054129-81b6c800-a3f7-11e9-9a5c-3dd20610ed40.png)

**Comment by lambourn** [comment]
@dmdnkv thanks for the heads-up. 

I just checked the Android App Bundle report for the uploaded .aab. It turns out that we'd miss 49 device models in total if we remove x86_64 and x86 support for our app. Looks acceptable.

**Comment by kzimny** [comment]
Just uploaded new version to google play with config recommended by @dmdnkv. I miss only five devices if x86 and x86_64 is removed, all of them are wearable devices (watches) which are not relevant for me. Thanks!

**Comment by jeremypele** [comment]
@rosen-vladimirov Do you guys plan to publish an official statement/blog post regarding this specific case? August 1st is in 2 weeks

**Comment by ExaltedJim** [comment]
> @ExaltedJim ,
> do you see any crash information in the logs?

@felix-idf sorry I have been busy as of late with other projects - when I find the time I will bring up my app in the 64 bit emulator again and report any crash logs.

Just to let others know, after compiling my app with 'armeabi-v7a', 'arm64-v8a', 'x86' ,'x86_64' enabled, I published the app to the store, and the previous warning I received from Google Play Store about lack of 64-bit support has gone away, however as I've demonstrated in the 64-bit emulator, it won't run.

**Comment by dmdnkv** [comment]
@ExaltedJim I had almost the same path, it was 32bit only, then I built with  'armeabi-v7a', 'arm64-v8a', 'x86' , 'x86_64' there were no warnings in the Google Console after upload it but then later after several weeks it was an email from Google on July 10th with a warning about the app didn't meet 64bit requirement.

Did you receive an email from Google?

**Comment by nlmunro** [comment]
I also got the warning email, despite adding the 64 bit options, so I did
as others suggested and removed 'x86' , 'x86_64'. For me that only excluded
Android Wear devices (which are not relevant to our app), so hopefully that
will solve it...

**Comment by ExaltedJim** [comment]
> @ExaltedJim I had almost the same path, it was 32bit only, then I built with 'armeabi-v7a', 'arm64-v8a', 'x86' , 'x86_64' there were no warnings in the Google Console after upload it but then later after several weeks it was an email from Google on July 10th with a warning about the app didn't meet 64bit requirement.
> 
> Did you receive an email from Google?

@dmdnkv I received the email before attempting to add 64-bit support to my app. I have not received an email since uploading my updated app (though my production app is still 32-bit only, the newly uploaded one is still in internal testing).

I had a look at the details of my 32-bit production app and it says that the native platforms it supports are "armeabi-v7a" and "x86".

When I look at the updated app in internal testing is says that is supports "all native platforms", which is encouraging. Additionally when I click on the button "Explore App Bundle" I see ABIs for "arm64-v8a", "armeabi-v7a", "x86_64" and "x86". 

@felix-idf - I just tried launching my app in debug mode on an Android x86_64 emulator and it crashes on starting up the app, with no information on why it is crashing. If you know where the logs might be I can grab them and display them here.

**Comment by ExaltedJim** [comment]
> @ExaltedJim ,
> do you see any crash information in the logs?

@felix-idf managed to work out how to get crash information, here is what I see in the logs:

`
2019-07-15 10:35:14.323 5826-5826/com.foo.Bar E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.foo.Bar, PID: 5826
    java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.foo.Bar-aUeAJgf5iB2UWUve7p6oEQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.foo.Bar-aUeAJgf5iB2UWUve7p6oEQ==/lib/x86_64, /data/app/com.com.Foo-aUeAJgf5iB2UWUve7p6oEQ==/base.apk!/lib/x86_64, /system/lib64]]] couldn't find "libNativeScript.so"
        at java.lang.Runtime.loadLibrary0(Runtime.java:1012)
        at java.lang.System.loadLibrary(System.java:1669)
        at com.tns.RuntimeHelper.initRuntime(RuntimeHelper.java:71)
        at com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:19)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1154)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5871)
        at android.app.ActivityThread.access$1100(ActivityThread.java:199)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1650)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
`

**Comment by racknoris** [comment]
@dmdnkv Did google support respond?

**Comment by dmdnkv** [comment]
@racknoris yes, they answered recently.
They didn't answer widely on my question but it has been said that the app, I am asking for to check after the last upload, has 64 bit support and I don't need to update anything regarding 64 bit case. So, I would say, that removing x86 platform from APK resolves the issue.

**Comment by tscislo** [comment]
Guys! I made the changes suggested by @dmdnkv and now I'm waiting for Google support to verify it. Is there and way to launch application with the following setup:
abiFilters.addAll(['armeabi-v7a','arm64-v8a', 'x86', 'x86_64'])
on Android emulator either 32 or 64 bit?

**Comment by dmdnkv** [comment]
@tscislo x86_64 fails on either real device or Android emulator, the reason is the lack of 'libNativeScript.so' in the output APK, but x86 (32bit) should work on Android emulator without any problems.

**Comment by tscislo** [comment]
@dmdnkv that is correct it works on emulator with x86, thanks!

**Comment by racknoris** [comment]
@tscislo did google respond?

**Comment by tscislo** [comment]
@racknoris yes, I did they simply said that if there is no more warning in Google Dev Console for this release it means that it is ok. Problem is that there has never been any warning, I just got an e-mail like all the others. So I assume it is done.

**Comment by racknoris** [comment]
I see, thank you

**Comment by mobilekbs** [comment]
> The same problem.
> It was decided to skip x86 platform support for our app as a workaround since almost 99% of mobile devices use ARM.
> The following config could help:
> 
> ```
>  defaultConfig {
>    generatedDensities = []
>    ndk {
>        abiFilters.clear()
>        abiFilters.addAll(['armeabi-v7a','arm64-v8a'])
>    }
>  }
>  aaptOptions {
>    additionalParameters "--no-version-vectors"
>  }
> }
> ```
> 
> Supported count of devices was decreased by 6 from 10557 to 10551 (that was displayed in Google Console App releases section after the roll out with that configuration)

Hello everyone,

I have the same problem. I am unable to get 64 bit libs to show up. I have followed these steps but still nothing. 

I even got the error:
> Cannot invoke method clear() on null object

My Gradle file:
       ```
        versionName "1.6.6"
        multiDexEnabled true
        vectorDrawables.useSupportLibrary = true

        ndk.abiFilters.clear()
        ndk.abiFilters.addAll(['armeabi-v7a', 'arm64-v8a'])

```
Can anyone guide me in a correct direction? 

Thank you

**Comment by nudiref** [comment]
any resolutions on this issue? i need to release a new version of my app and just encountered it. i tried all the solutions suggested here to no avail. 

when i install the app on my device, like @ExaltedJim reports, it crashes upon launch. i get a blank white screen and it stops working. i can never see the initial screen

i tired adding following to the app.gradle:

`    ndk {
       abiFilters.clear()
       abiFilters.addAll(['armeabi-v7a','arm64-v8a', 'x86', 'x86_64'])
   }`

and also:

`    ndk {
       abiFilters.clear()
       abiFilters.addAll(['armeabi-v7a','arm64-v8a', 'x86'])
   }`

and

`    ndk {
       abiFilters.clear()
       abiFilters.addAll(['armeabi-v7a','arm64-v8a'])
   }`

without 'x86' and/or 'x86_64'.

**Comment by dmdnkv** [comment]
@nudiref I would suggest you to check the logs and see what is the reason of the crash and if you see something like ```couldn't find "libNativeScript.so"``` it is related to this x86_64 issue, otherwise it is something else. 

If you run debug build on the device it is also might be worthy to do a clean rebuild after those gradle manipulations:
```tns platform clean android```
```tns build android --clean```

Moreover, if you said that you see a blank screen it might be not x86_64 problem because when I had this problem it is even didn't launch anything and showed the app crash message on the OS home page.
I think the reason of that is this missing libNativeScript.so that might be required to at least start to show something after it has been loaded.

**Comment by nudiref** [comment]
thanks for the feedback @dmdnkv 

when i look at the logcat i see the following:

`28772 F DEBUG   : backtrace:`
`28772 F DEBUG   :     #00 pc 0000000000b03ecc  /data/app/com.myApp-NUC4XcsB72yaChZyxNYiTg==/lib/arm64/libNativeScript.so`
`28772 F DEBUG   :     #01 pc 0000000000b04e74  /data/app/com.myApp-NUC4XcsB72yaChZyxNYiTg==/lib/arm64/libNativeScript.so`
`28772 F DEBUG   :     #02 pc 0000000000ad9870  /data/app/com.myApp-NUC4XcsB72yaChZyxNYiTg==/lib/arm64/libNativeScript.so (__cxa_throw+112)`
`28772 F DEBUG   :     #03 pc 00000000001c75d0  /data/app/com.myApp-NUC4XcsB72yaChZyxNYiTg==/lib/arm64/libNativeScript.so (SIGABRT_handler(int)+132)`

does that mean that the problem is due to 'arm64-v8a'?

**Comment by dmdnkv** [comment]
@nudiref according to my understanding of the log you provided I could assume that libNativeScript.so is there in place, since the trace shows that some execution is going on inside it, so it seems to be the problem isn't related to its absence that is what we are discussing in this issue.

I would suggest trying to debug it might be by excluding some libraries/plugins one by one or whatever you use in your app  until it works. Did you try to run a blank "Hello World" app on that device with the same build options?

**Comment by n0mer** [comment]
Today Google Play Store  stopped accepting APKs without 64-bit support.

And, this page https://docs.nativescript.org/core-concepts/android-runtime/requirements does not have `arm64-v8a` in the list of supported ABIs

**Comment by FanniPutnoczki** [comment]
So we cannot publish apps to the Google Play anymore? I just wanted publish it today and I get the same error message. I didn't get it yesterday

**Comment by kzimny** [comment]
change your app.gradle as follow:
```
android {
  defaultConfig {
    generatedDensities = []
	  applicationId = 'org.test.myapp'
    ndk {
      abiFilters.clear()
      abiFilters.addAll(['armeabi-v7a','arm64-v8a'])
    }
  }
  aaptOptions {
    additionalParameters '--no-version-vectors'
  }
}
```
It works! Look at my previous [post](https://github.com/NativeScript/nativescript-cli/issues/4330#issuecomment-510534508).

**Comment by n0mer** [comment]
@kzimny the point is that with such config it does not.

```
android {
    defaultConfig {
        minSdkVersion 17
        multiDexEnabled true
        generatedDensities = []
        ndk.abiFilters.clear()
        ndk.abiFilters.addAll([ 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'])
    }
    aaptOptions {
        additionalParameters "--no-version-vectors"
    }
}
```

Can you please give any hint how to debug this, or make it work?

**Comment by kzimny** [comment]
Why do you need x86 and x86_64 devices? Please read my prevoius posts:
1. https://github.com/NativeScript/nativescript-cli/issues/4330#issuecomment-510478500
2. https://github.com/NativeScript/nativescript-cli/issues/4330#issuecomment-510534508
Believe, you don't need x86 and x86_64 devices except you app should run on watches wearable devices.

**Comment by NathanaelA** [comment]
@kzimny - Actually their are some x86 devices, Asus made several of them (in fact I have access to one of them for testing.  :grinning:)    that are decent devices.   So removing x86 will eliminate these devices from using the app.

**Comment by kzimny** [comment]
Check how many devices you will loose and decide yourself what is better. Wait for the next release 6.x.x or upgrade as described above.

**Comment by NathanaelA** [comment]
For those who care; I can confirm that because NS doesn't ship with the X86-64 bit runtime, you are now blocked on uploading UNLESS you strip out all x86 support (see Kzimny's post above https://github.com/NativeScript/nativescript-cli/issues/4330#issuecomment-522721510). 

You will see this message:
![image](https://user-images.githubusercontent.com/850871/63303405-5e5c2400-c2a5-11e9-9838-de6202240ca5.png)

Which links you to this document: 
https://developer.android.com/distribute/best-practices/develop/64-bit

Which specifically does say that you need a 64bit runtime for x86 to deploy.

**Comment by FanniPutnoczki** [comment]
I ended up excluding the whole x86 architecture and using only the arm the same way @kzimny did, and I lost only 5 devices that needed it anyway, so it's not a big loss at all :)

**Comment by NathanaelA** [comment]
@FanniPutnoczki - Just a FYI; those 5 devices probably include a several ASUS phones where are very popular in some areas.    So once NS fixes the issue; I would highly recommend you do release a version for the x86 market...  :D

**Comment by edrick27** [comment]
![image](https://user-images.githubusercontent.com/20235620/63313541-14356b80-c2c2-11e9-9f0b-168dcde7e7e9.png)

I have the same problem, but if excluding the whole x86 architecture, I will lose a lot devices 😢

**Comment by n0mer** [comment]
@kzimny just FYI - confirm that your config worked.

Also, quote from https://developer.android.com/distribute/best-practices/develop/64-bit#multi-apk-compliance :


> If one of your APKs is marked as not being compliant, but is older and it’s not possible to bring it into compliance, one strategy is to add a maxSdkVersion="27" attribute in the [uses-sdk](https://developer.android.com/guide/topics/manifest/uses-sdk-element) element in that APK’s manifest. This APK won’t be delivered to devices running Android 9 Pie or later, and it will no longer block compliance


If maxSdkVersion in AndroidManifest.xml is honored by {N}/webpack build (in the light of https://github.com/NativeScript/android-runtime/issues/575), then this can serve as another work-around.

**Comment by rosen-vladimirov** [comment]
Hey everyone,
It looks like Google have changed the requirements in the last few days - we've verified we can publish applications with 'armeabi-v7a', 'arm64-v8a', 'x86' architectures around August 1st (when the requirement was set) and we did not have any troubles. But it turns out something is changed now.

Due to the change in the requirements we are preparing a patch release for Android runtime (tns-android) that will include x86_64 architecture as well. Stay tuned for more information and thank you all for sharing the details about the errors.

**Comment by n0mer** [comment]
Just in case - one more thing.

![image](https://user-images.githubusercontent.com/1862997/63337753-d11add00-c341-11e9-9034-052bddef8c56.png)

For local debug purposes in Android emulator, only arm AVDs can be used when building for `'armeabi-v7a', 'arm64-v8a'`. For now, android studio provides arm AVDs up to android 7.1.1. So, no android 8 or 8, and they are lack Intel's HAXM h/w acceleration: https://developer.android.com/studio/run/emulator-acceleration

Hope this helps :)

**Comment by Roar1827** [comment]
@rosen-vladimirov looks like you made some big progress on this today- any guesses at when this issue will be resolved? Thank you very much for your efforts, we greatly appreciate them!

**Comment by n0mer** [comment]
@Roar1827 this might help for now: https://github.com/NativeScript/NativeScript/issues/7705#issuecomment-523363425

**Comment by Roar1827** [comment]
Thanks @n0mer . We were able to get a build up using that. Looks like we've run into some Android compatibility warnings in the Pre-launch report- not sure how serious they are, but we will do some testing. Thank you for your help!

**Comment by rosen-vladimirov** [comment]
Hey everyone,
As you've noticed - we've made a significant progress on this issue and we expect to release a patch version tomorrow (August 22nd) in case our tests show everything is working as expected. Meanwhile, in case you have some urgent release, you can use the `rc` versions of nativescript-dev-webpack and `tns-android` packages. I.e.:
```
rm -rf platforms
npm i --save-dev nativescript-dev-webpack@rc
tns platform add android@rc
tns build android --release --env.snapshot...
```
Stay tuned for more information, hopefully we'll be able to release a fix as planned.

**Comment by vinita1505** [comment]
Hi everyone,

If project doesn’t contain any native library or I can say when I analyze the apk I dint got any .so file. So Is my apk or an app is compatible for 64 bit?

**Comment by felix-idf** [comment]
@vinita1505 All NativeScript apps for Android are shipped with the V8 JavaScript-engine, which is native code, so any NativeScript developer is affected by this issue.

But the NS team has just released tns-android 6.0.2, so update your project and you should be safe now.

**Comment by jnorkus** [comment]
> But the NS team has just released tns-android 6.0.2, so update your project and you should be safe now.

Is it safe to upgrade platform modules without upgrading to NS6?

**Comment by felix-idf** [comment]
@jnorkus no, you have to migrate your project to NS 6 to be able to use the latest Android runtime for your app.

**Comment by rosen-vladimirov** [comment]
Hey everyone,
We've just released `tns-android` 6.0.2 and `nativescript-dev-webpack` 1.1.1.
You need to update both of them (and you need to target NativeScript 6):
```
rm -rf platforms
npm i --save-dev nativescript-dev-webpack@1.1.1
tns platform add android@6.0.2
tns build android --release --env.snapshot..
```

**Comment by jnorkus** [comment]
Is there any workaround for NS5 users that includes x86 support? Many plugins out there don't support NS6 and most of us are stuck with that.

**Comment by etabakov** [comment]
@jnorkus we don't plan to bakport this fix to older version as it is not a trivial amount of work. We would rather put effort to support the users in migrating to 6.0. Can you send us a summary of the plugins that incompatible, so that we can try to help where we can? Does this approach sound reasonable to you?

**Comment by jnorkus** [comment]
@etabakov I can totally understand the reasoning. To be honest I haven't even tried migrating yet. I am assuming it won't work because of the AndroidX breaking change. Once I have time to look into migration I will report my progress.

For the time being I will just exclude x86 support for our apps.

**Comment by marast78** [comment]
> Just uploaded new version to google play with config recommended by @dmdnkv. I miss only five devices if x86 and x86_64 is removed, all of them are wearable devices (watches) which are not relevant for me. Thanks!

@kzimny You need those (x86 and x86_64) mainly to test in emulators

**Comment by NathanaelA** [comment]
@marast78 
If you are continuing to use NS 4/5 --   (Again this is NOT needed anymore with 6.02+)
```
android {
  defaultConfig {
    minSdkVersion 23
    generatedDensities = []
    ndk {
      abiFilters.clear()
      abiFilters.addAll(['armeabi-v7a','arm64-v8a'])
    }
  }    
  buildTypes {
    debug {
      ndk {
        abiFilters 'x86'
      }
    }
  }
  aaptOptions {
    additionalParameters "--no-version-vectors"
  }
}
```
This will disable x86 for release builds; but enable x86 for debug builds --- Thanks to Peter Staev who posted it in slack...

**Comment by marast78** [comment]
@NathanaelA awesome tip :+1: thanks @PeterStaev

**Comment by kefahB** [comment]
Hi,

I use currently tns `5.4` and have same issu, I trying to severals All solutions here but the Play s-Store still give me same error !

**Comment by kefahB** [comment]
In fact what ever changes I do on `app.gradle` on `platforms/android/app/build.gradle` on `defaultConfig` section is same : 

```ts
defaultConfig {
        def manifest = new XmlSlurper().parse(file(android.sourceSets.main.manifest.srcFile))
        def minSdkVer = manifest."uses-sdk"."@android:minSdkVersion".text() ?: 17
        minSdkVersion minSdkVer
        targetSdkVersion computeTargetSdkVersion()
        ndk {
            if (onlyX86) {
                abiFilters 'x86'
            } else {
                abiFilters 'x86', 'armeabi-v7a', 'arm64-v8a'
            }
        }
        dexOptions {
            jumboMode = true
        }
    }
```

**Comment by skillbrighterinfotech** [comment]
> I just built an app with `ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86'`, exported as .aab. and uploaded it to the Play Store just fine, no complaints about missing 64-bit binaries.
> 
> My assumption is that Google does not require x86_64 binaries at all (and if the app is really running on a x86_64 target, e.g. tablet. then the x86 32-bit binaries will be used)

I have do this but not work same issue

**Comment by skillbrighterinfotech** [comment]
> I added this to my app.gradle and it worked fine.
> 
> ```
> defaultConfig {
>        ......
>        ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86','x86_64'
> }
> ```

Not work , please help me

**Comment by n0mer** [comment]
@Clifte please try "clear" first

```
        ndk.abiFilters.clear()
        ndk.abiFilters.addAll([ 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'])
```

**Comment by sebj54** [comment]
@NathanaelA Thank you for the snippet! I also wanted to clarify, the `minSdkVersion` is not mandatory. I removed it because I still have users with Android < 6 and I was able to publish on Google Play 👌

**Comment by nudiref** [comment]
hello all,

what worked for me in the end was the following configuration:

`android {  `
`  defaultConfig {   `
`    ........   `
`    ndk {`
`       abiFilters.clear()`
`       abiFilters.addAll(['armeabi-v7a', 'x86','x86_64'])`
`   } `
`   .......`
`  }  `
`} `

- when i include 'arm64-v8a', app will build but it will not launch (crashes as soon as it is launched)
- i am still using nativescript version 3.4.0. i am not able to upgrade due to many errors upgrade entails.
- app generated with the configuration above will not work in devices with android version < 9, because of this, i had to use multi-apk feature of android. i created one apk with above configuration for minSdkVersion = 28, and another with no ndk node in app,gradle with maxSdkVersion = 27. with these two, android accepted my release. by having an apk with above configuration was enough to satisfy 64 bit compliance requirement, so, it seems it is enough to include one of 'arm64-v8a' or 'x86_64' to satisfy google's requirement. 

i would love to know what is wrong with 'arm64-v8a' package and be able to include it.

**Comment by Izrab** [comment]
I use NativeScript 6.1 
Do I need to do anything regarding the 64-bit requirement issue?

**Comment by sebj54** [comment]
I have NS 6.1.2 on a fresh new project and I still cannot upload an `aab` file to Google Play.

The bundle has been generated with this command:

```
tns build android --bundle --release --key-store-path upload-key.keystore --key-store-password *** --key-store-alias upload-key --key-store-alias-password *** --aab
```

Here is my app.gradle:

```
// Add your native dependencies here:

android {
  defaultConfig {
    generatedDensities = []
    applicationId = "com.example.app"
  }
  aaptOptions {
    additionalParameters "--no-version-vectors"
  }
}
```

What did I miss?

**Comment by racknoris** [comment]
@sebj54 Did you manage to solve this?

**Comment by sebj54** [comment]
@racknoris I'm still using this suggestion: https://github.com/NativeScript/nativescript-cli/issues/4330#issuecomment-523993414, even with NS 6

**Comment by racknoris** [comment]
@sebj54 I see, thank you

**Comment by Fatme** [comment]
@sebj54,

It should be enough just executing the command you mentioned:
```
tns build android --bundle --release --key-store-path upload-key.keystore --key-store-password *** --key-store-alias upload-key --key-store-alias-password *** --aa
```

No additional actions are required.

However, fell free to open an another issue with detailed description and exact error you received when trying to upload an `aab` file to `Google Play`.
